### DISCUSSION TITLE
Smart Zigbee 4 Gang Switch support problem

### DISCUSSION THREAD

**Discussion by Martin-Dolan** [Koenkk/zigbee2mqtt#17162] [Devices]
I've purchased a few of the MakeGood 4gang switches, and I can get them to pair, however the show as unsupported. 

I've tried for 2 days top write a custom converter, but I'm not a programmer and that didn't work . got to the point where in the end my Zigbee2MQTT wouldn't even load and I had to restore from a backup haha
Tried the Home assistant community and that didn't help.

model: TS0601
Zigbee Manufacturer:  _TZE200_mexisfik

This is where I purchased them from:
https://smartguys.com.au/makegood-zigbee-4-gang-smart-light-switch.html

Blakadder says they are supported (but cant get them to work and I've searched for the model /manufacturer string and it doesn't show in zigbee2MQTT:
https://zigbee.blakadder.com/MakeGood_MG-ZG04W.html

This is a page where they were added to ZHA if that helps:
https://github.com/zigpy/zha-device-handlers/issues/1634

Can anyone help?  It's not like I haven't given it my best go..  

NOTE:  to get one to pair, all the documentation says just hold a button for 10 seconds.  That doesn't work, you have to turn one of the switches on/of about 5 times and on the 5th time you turn it on, keep your finger on the button for about 5 seconds until they all start to flash.  Then let go and it will pair.   Wasted about an hour just working that out :/

**Comment by Koenkk** [comment]
could you provide a screenshot of the device -> about page?

**Comment by Martin-Dolan** [comment]
sure, this is after removing my failed attempt of a custom .js file haha

![about](https://user-images.githubusercontent.com/11151228/228238609-5300762c-17a1-4a8e-a046-2153333d34e2.jpg)

**Comment by Martin-Dolan** [comment]
Just checking that was what you need, let me know if you need anything else.

**Comment by Martin-Dolan** [comment]
Just wondering if I should install the edge version of Zigbee2MQTT on home assistant?  Would that help test any changes you make to get this working?

**Comment by Koenkk** [comment]
No need to switch to edge, please try with this converter:

```js
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const ota = require('zigbee-herdsman-converters/lib/ota');
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const utils = require('zigbee-herdsman-converters/lib/utils');
const globalStore = require('zigbee-herdsman-converters/lib/store');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    fingerprint: tuya.fingerprint('TS0014', ['_TZ3000_jr2atpww', '_TYZB01_dvakyzhd', '_TZ3000_mrduubod',
        '_TZ3210_w3hl6rao', '_TYZB01_bagt1e4o', '_TZ3000_r0pmi2p3', '_TZ3000_fxjdcikv', '_TZ3000_q6vxaod1', '_TZE200_mexisfik']),
    model: 'TS0014',
    vendor: 'TuYa',
    description: 'Smart light switch - 4 gang without neutral wire',
    extend: tuya.extend.switch({backlightModeLowMediumHigh: true, endpoints: ['l1', 'l2', 'l3', 'l4']}),
    endpoint: (device) => {
        return {'l1': 1, 'l2': 2, 'l3': 3, 'l4': 4};
    },
    whiteLabel: [{vendor: 'TUYATEC', model: 'GDKES-04TZXD'}, {vendor: 'Vizo', model: 'VZ-222S'},
        {vendor: 'MakeGood', model: 'MG-ZG04W/B/G'}, {vendor: 'Mercator Ikuü', model: 'SSW04'}],
    meta: {multiEndpoint: true},
    configure: async (device, coordinatorEndpoint, logger) => {
        await tuya.configureMagicPacket(device, coordinatorEndpoint, logger);
        try {
            for (const ID of [1, 2, 3, 4]) {
                const endpoint = device.getEndpoint(ID);
                await reporting.bind(endpoint, coordinatorEndpoint, ['genOnOff']);
            }
        } catch (e) {
            // Fails for some: https://github.com/Koenkk/zigbee2mqtt/issues/4872
        }
        device.powerSource = 'Mains (single phase)';
        device.save();
    },
};

module.exports = definition;

```

**Comment by Martin-Dolan** [comment]
Oh wow,  That's the converter I found on the web and tried to modify it to get it to work before I asked you for help!   No luck.

Basically it wont recognise the device unless you change the fingerprint from
`    fingerprint: tuya.fingerprint('TS0014', ['_TZ3000_jr2atpww', '_TYZB01_dvakyzhd', '_TZ3000_mrduubod',
        '_TZ3210_w3hl6rao', '_TYZB01_bagt1e4o', '_TZ3000_r0pmi2p3', '_TZ3000_fxjdcikv', '_TZ3000_q6vxaod1', '_TZE200_mexisfik']),`

to
 `    fingerprint: tuya.fingerprint('TS0601', ['_TZ3000_jr2atpww', '_TYZB01_dvakyzhd', '_TZ3000_mrduubod',
        '_TZ3210_w3hl6rao', '_TYZB01_bagt1e4o', '_TZ3000_r0pmi2p3', '_TZ3000_fxjdcikv', '_TZ3000_q6vxaod1', '_TZE200_mexisfik']),`

as it needs the 'TS0601'

(I know that's probably going to break the other devices listed but I'm getting to the end of my ability here!

Then if I go to the device  -> exposes -> and click the on/off buttons on the endpoints I get a timeout error on l1 and a "device '4gang' has no endpoint (l2, l3, l4 etc) " on the others

![errors](https://user-images.githubusercontent.com/11151228/229294284-b10bc5f9-f0c3-4556-9d05-c45fbdabba13.jpg)

**Reply by Koenkk** [reply]
Try with:

```js
const fz = require('zigbee-herdsman-converters/converters/fromZigbee');
const tz = require('zigbee-herdsman-converters/converters/toZigbee');
const exposes = require('zigbee-herdsman-converters/lib/exposes');
const reporting = require('zigbee-herdsman-converters/lib/reporting');
const extend = require('zigbee-herdsman-converters/lib/extend');
const ota = require('zigbee-herdsman-converters/lib/ota');
const tuya = require('zigbee-herdsman-converters/lib/tuya');
const utils = require('zigbee-herdsman-converters/lib/utils');
const globalStore = require('zigbee-herdsman-converters/lib/store');
const e = exposes.presets;
const ea = exposes.access;

const definition = {
    fingerprint: [{modelID: 'TS0601', manufacturerName: '_TZE200_aqnazj70'}, {modelID: 'TS0601', manufacturerName: '_TZE200_k6jhsr0q'},
        {modelID: 'TS0601', manufacturerName: '_TZE200_di3tfv5b'}, {modelID: 'TS0601', manufacturerName: '_TZE200_mexisfik'}],
    model: 'TS0601_switch_4_gang',
    vendor: 'TuYa',
    description: '4 gang switch',
    exposes: [e.switch().withEndpoint('l1').setAccess('state', ea.STATE_SET),
        e.switch().withEndpoint('l2').setAccess('state', ea.STATE_SET),
        e.switch().withEndpoint('l3').setAccess('state', ea.STATE_SET),
        e.switch().withEndpoint('l4').setAccess('state', ea.STATE_SET)],
    fromZigbee: [fz.ignore_basic_report, fz.tuya_switch],
    toZigbee: [tz.tuya_switch_state],
    meta: {multiEndpoint: true},
    endpoint: (device) => {
        // Endpoint selection is made in tuya_switch_state
        return {'l1': 1, 'l2': 1, 'l3': 1, 'l4': 1};
    },
};

module.exports = definition;

```

**Reply by Martin-Dolan** [reply]
You nailed it!!  
That's awesome.  No errors.
Switch works fine, Home assistant control works fine.

Thankyou so much .  I'm trying to get my head around making these so I can give back to the community.  I was looking for an existing one that was based on TS0601 , but couldn't find it.  Now I can use this to search around and learn more.

**Reply by Koenkk** [reply]
Added!

Changes will be available in the dev branch in a few hours from now. (https://www.zigbee2mqtt.io/advanced/more/switch-to-dev-branch.html)

**Comment by Martin-Dolan** [comment]
❤️
